juju/juju
# feat: add controller model ID to ServerConfig, sharedServerContext

This PR fixes a chicken-and-egg problem in setting up the apiserver. Previously we were given just a service factory getter. We needed to pass in the controller model UUID to obtain the service factory (for the controller model). However, we needed the service factory to get the controller model UUID.

To get around this, we were using a nasty hack where we passed in the controller namespace (not UUID). This worked because the controller service factory methods don't depend on the UUID passed in. (In fact, any string would have worked). Then we could obtain the controller UUID from this "false" service factory, and pass it back in to the service factory getter to get the "real" service factory. Something like this:
```go
// the arg here should be a model UUID, but we don't have this yet
fakeServiceFactory := serviceFactoryGetter.FactoryForModel(database.ControllerNS)
controllerModel, err := fakeServiceFactory.Model().ControllerModel(ctx)
if err != nil { ... }
realServiceFactory := serviceFactoryGetter.FactoryForModel(controllerModel.UUID.String())
...
```

To fix this, we add the `controllerModelID` as a field to `apiserver.ServerConfig` and `api.sharedServerContext`. Then, we will always have the controller model UUID when we need it to pass in to the service factory getter.

To nudge developers in the right direction, I've changed the arg type of `ServiceFactoryGetter.FactoryForModel` to `core/model.UUID`:
```go
type ServiceFactoryGetter interface {
	// FactoryForModel returns a ServiceFactory for the given model.
	FactoryForModel(modelID model.UUID) ServiceFactory
}
```
This should reduce the temptation to pass in a namespace or fake value that is not a valid model UUID.

Drive-bys:
- move apiserver worker services to `service.go`
- change some instances of `modelUUID string` to `modelID model.UUID`

## Checklist




## QA steps



Bootstrap Juju and run `smoke` tests.

Check that the GitHub actions pass below.

## Links



**Jira card:** JUJU-6323
